src/analysis/processing/qgsalgorithmxyztiles.cpp

```diff
@@ -0,0 +1,511 @@
+    }
+    QTextStream fout( &htmlFile );
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    fout.setCodec( "UTF-8" );
+#endif
+    fout << html;
+
+    results.insert( QStringLiteral( "OUTPUT_HTML" ), outputHtml );
+  }
+
+  return results;
+}
+
+void QgsXyzTilesDirectoryAlgorithm::processMetaTile( QgsMapRendererSequentialJob *job )
+{
+  MetaTile metaTile = mRendererJobs.value( job );
+  QImage img = job->renderedImage();
+
+  QMap<QPair<int, int>, Tile>::const_iterator it = metaTile.tiles.constBegin();
+  while ( it != metaTile.tiles.constEnd() )
+  {
+    QPair<int, int> tm = it.key();
+    Tile tile = it.value();
+    QImage tileImage = img.copy( mTileWidth * tm.first, mTileHeight * tm.second, mTileWidth, mTileHeight );
+    QDir tileDir( QStringLiteral( "%1/%2/%3" ).arg( mOutputDir ).arg( tile.z ).arg( tile.x ) );
+    tileDir.mkpath( tileDir.absolutePath() );
+    int y = tile.y;
+    if ( mTms )
+    {
+      y = tile2tms( y, tile.z );
+    }
+    tileImage.save( QStringLiteral( "%1/%2.%3" ).arg( tileDir.absolutePath() ).arg( y ).arg( mTileFormat.toLower() ), mTileFormat.toStdString().c_str(), mJpgQuality );
+    ++it;
+  }
+
+  mRendererJobs.remove( job );
+  job->deleteLater();
+
+  mFeedback->setProgress( 100.0 * ( mProcessedTiles++ ) / mTotalTiles );
+
+  if ( mMetaTiles.size() > 0 )
+  {
+    startJobs();
+  }
+  else if ( mMetaTiles.size() == 0 && mRendererJobs.size() == 0 )
+  {
+    mEventLoop->exit();
+  }
+}
+
+// Native XYZ tiles (MBTiles) algorithm
+
+QString QgsXyzTilesMbtilesAlgorithm::name() const
+{
+  return QStringLiteral( "tilesxyzmbtiles" );
+}
+
+QString QgsXyzTilesMbtilesAlgorithm::displayName() const
+{
+  return QObject::tr( "Generate XYZ tiles (MBTiles)" );
+}
+
+QStringList QgsXyzTilesMbtilesAlgorithm::tags() const
+{
+  return QObject::tr( "tiles,xyz,tms,mbtiles" ).split( ',' );
+}
+
+QString QgsXyzTilesMbtilesAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Generates XYZ tiles of map canvas content and saves them as an MBTiles file." );
+}
+
+QgsXyzTilesMbtilesAlgorithm *QgsXyzTilesMbtilesAlgorithm::createInstance() const
+{
+  return new QgsXyzTilesMbtilesAlgorithm();
+}
+
+void QgsXyzTilesMbtilesAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  createCommonParameters();
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT_FILE" ), QObject::tr( "Output" ), QObject::tr( "MBTiles files (*.mbtiles *.MBTILES)" ) ) );
+}
+
+QVariantMap QgsXyzTilesMbtilesAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString outputFile = parameterAsString( parameters, QStringLiteral( "OUTPUT_FILE" ), context );
+
+  QgsProject *project = context.project();
+
+  QgsCoordinateReferenceSystem wgsCrs( "EPSG:4326" );
+  QgsCoordinateReferenceSystem mercatorCrs( "EPSG:3857" );
+  QgsCoordinateTransform src2wgs( project->crs(), wgsCrs, context.transformContext() );
```

Comment:
I think that should be moved to `prepare`.

Reply:
Thanks, done now.